errors.go

```diff
@@ -0,0 +1,164 @@
+package jsonapi
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+)
+
+// MarshalErrors writes a JSON API response using the given `[]error`.
+// This function works by passing each `error` in the given slice through a routine which attempts
+// to cast the error to each of this package's `Error<FieldName>Compatible` interfaces. Though
+// it works for errors which do not implement any of these interfaces as well. For more information
+// on JSON API error payloads, see the spec here: http://jsonapi.org/format/#document-top-level
+// and here: http://jsonapi.org/format/#error-objects
+//
+// To use this function in your code effectively, make your package's error types implement some
+// or all of the error compatibility interfaces (`ErrorTitleCompatible`, `ErrorDetailCompatible`,
+// ...). Doing so gives you direct control over what will appear in the serialied error object's
+// fields. Without implementing these interfaces, the `title` & `detail` fields will be populated
+// with derived data about the error.
+func MarshalErrors(w io.Writer, errs []error) error {
+	// Serialize the given errors.
+	var formattedErrors []ErrorObject
+	for _, err := range errs {
+		e := marshalError(err)
+		formattedErrors = append(formattedErrors, e)
+	}
+
+	// Write out the serialize errors payload.
+	if err := json.NewEncoder(w).Encode(&ErrorsPayload{Errors: formattedErrors}); err != nil {
+		return err
+	}
+	return nil
+}
+
+// marshalError will serialize the given error as best as possible according to this
+// package's `Error<field>Compatible` interfaces.
+func marshalError(err error) ErrorObject {
+	errorObject := ErrorObject{}
+	if e, ok := err.(ErrorIDCompatible); ok {
+		errorObject.ID = e.GetID()
+	}
+
+	if e, ok := err.(ErrorTitleCompatible); ok {
+		errorObject.Title = e.GetTitle()
+	} else {
+		errorObject.Title = fmt.Sprintf("Encountered error of type: %T", err)
```

Comment:
I see your point. And I can agree with it to a certain extent. The issue then becomes: `MarshalErrors` doesn't really work with a true `[]error`. Because if one of there `error`s in the slice does not implement any of the interfaces defined here, you will get an empty JSON object in the response `errors` array. Which would be completely meaningless. Why fall back to that instead of have a sane default which works with everything?

In your example, the app could still choose to use only the `code` field. If they are **actually serious** about reducing bytes on the wire, they should be using gRPC.